Email table new columns

alfredoct96 opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | features
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | N/A

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "features" branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:

Added three new columns in email table:

* created at
* modified at
* created by

### Before
![mauticEmailTableBefore](https://user-images.githubusercontent.com/50916237/108881366-839cc180-7603-11eb-9be7-ed4aa21fbb0b.png)

### After
![mauticEmailTableAfter](https://user-images.githubusercontent.com/50916237/108881371-8697b200-7603-11eb-99f9-4d0b27ddf4bf.png)


<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create an email
3. Check existence of new columns

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9685"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

alfredoct96 committed:
Added the ordered columns creation date and modified date

alfredoct96 committed:
Moved ID column to the end

alfredoct96 committed:
Added created by column

escopecz commented on app/bundles/EmailBundle/Views/Email/list.html.php:
```diff
@@ -217,6 +250,9 @@
                         <?php echo $view['content']->getCustomContent('email.stats.below', $mauticTemplateVars); ?>
                     </td>
+                    <td class="visible-md visible-lg"><?php echo $item->getDateAdded() ? $item->getDateAdded()->format('Y-m-d H:i:s') : ''; ?></td>
+                    <td class="visible-md visible-lg"><?php echo $item->getDateModified() ? $item->getDateModified()->format('Y-m-d H:i:s') : ''; ?></td>
```
Use the formats that users can configure like
```suggestion
                    <td class="visible-md visible-lg"><?php echo $item->getDateModified() ? $view['date']->toFull($item->getDateModified()) : ''; ?></td>
```

alfredoct96 replied:
DoneHi @escopecz !

I added the two screenshot.

Removed visible-md is a good idea, i changed it and the method to render the date!

escopecz reviewed (changes requested):
@alfredoct96 can you please add a screenshot of the table before and after your change and add it to the PR description so the community can easily decide whether they want this UI change?

I'm worried it will be cluttered on smaller screens. Maybe we should remove the `visible-md` and leave those columns visible only for `visible-lg`. What do you think?

alfredoct96 committed:
Changed method for date

alfredoct96 committed:
Removed visible-md

alfredoct96 commented:
> @alfredoct96 can you please add a screenshot of the table before and after your change and add it to the PR description so the community can easily decide whether they want this UI change?
> 
> I'm worried it will be cluttered on smaller screens. Maybe we should remove the `visible-md` and leave those columns visible only for `visible-lg`. What do you think?

Hi @escopecz !

I added the two screenshot.

Removed visible-md is a good idea, i changed it and changed method to render date!

alfredoct96 committed:
Added test for index action

alfredoct96 committed:
CreatedBy is visible only on lg

escopecz reviewed (dismissed):
Code-wise this PR is good to go. It's not fetching any new data. The data this PR displays are already fetched form the database but just not used so far. 👍 

I'm far from expert on UI/UX so I'd let someone with artistic sense to approve that part and do a manual test.

kuzmany commented on app/bundles/EmailBundle/Views/Email/list.html.php:
```diff
@@ -217,6 +250,9 @@
                         <?php echo $view['content']->getCustomContent('email.stats', $mauticTemplateVars); ?>
                         <?php echo $view['content']->getCustomContent('email.stats.below', $mauticTemplateVars); ?>
                     </td>
+                    <td class="visible-lg"><?php echo $item->getDateAdded() ? $view['date']->toFull($item->getDateAdded()) : ''; ?></td>
```
Better use toFullConcat method (https://github.com/mautic/mautic/blob/features/app/bundles/CoreBundle/Templating/Helper/DateHelper.php#L111) which make shorten date time without timezone on the end

alfredoct96 replied:
Yes, use toFullConcat is a better idea, i replaced toFull with it

kuzmany reviewed (changes requested):
In comment I recommend us toFullConcat method for date.

From my side I don't like this kind of complex list, I prefer the simple one. Maybe should be better make it configurable same like we do for contacts columns. Maybe configurable in general for all lists. What do you think?

![image](https://user-images.githubusercontent.com/462477/109419652-d0e7ad00-79ce-11eb-85df-471ba6c5a602.png)

But If community vote for it, I cannot block it.

alfredoct96 committed:
Replace toFull with toFullConcat

RCheesley commented:
> Maybe should be better make it configurable same like we do for contacts columns. Maybe configurable in general for all lists. What do you think?

This was my thought as well - I think it would be nicer to have this as an option to enable. My concern with that is the way we have it implemented I think the config file will become super long with all the additional fields being specified if we take the same approach!  Is there a more elegant way to save the columns maybe?

dennisameling commented:
Hi @alfredoct96! Thanks for your PR 🚀 we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since we don't have access to your branch, could you please click "Update branch" at the bottom of this PR? If that doesn't work, you can also run `git merge upstream/features` locally. Thanks in advance!

![image](https://user-images.githubusercontent.com/17739158/112621033-90dae500-8e29-11eb-8541-bd0a5d36c2ba.png)

alfredoct96 commented:
> Hi @alfredoct96! Thanks for your PR  we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since we don't have access to your branch, could you please click "Update branch" at the bottom of this PR? If that doesn't work, you can also run `git merge upstream/features` locally. Thanks in advance!
> 
> ![image](https://user-images.githubusercontent.com/17739158/112621033-90dae500-8e29-11eb-8541-bd0a5d36c2ba.png)

Done @dennisameling

dennisameling commented:
>Maybe should be better make it configurable same like we do for contacts columns. Maybe configurable in general for all lists. What do you think?

Making it configurable would be great indeed. I wrote a plugin for a customer and did exactly that (very basic):

https://user-images.githubusercontent.com/17739158/121674251-19286800-cab2-11eb-88c4-fa6582e81a86.mp4

I can't make the plugin code available since it's way too customized for that customer, but can definitely create a minimal plugin with that specific logic 👍🏼 let me know if the team wants me to do that.

Would be best to store the settings on the user-level IMO, as different users might have different needs for their screens

escopecz commented on app/bundles/EmailBundle/Views/Email/list.html.php:
```diff
@@ -217,6 +250,9 @@
                     </td>
+                    <td class="visible-lg"><?php echo $item->getDateAdded() ? $view['date']->toFullConcat($item->getDateAdded()) : ''; ?></td>
+                    <td class="visible-lg"><?php echo $item->getDateModified() ? $view['date']->toFullConcat($item->getDateModified()) : ''; ?></td>
+                    <td class="visible-lg"><?php echo $item->getCreatedByUser(); ?></td>
```
There is a chance for XSS. Let's not introduce new vulnerabilities.
```suggestion
                    <td class="visible-lg"><?php echo $view->escape($item->getCreatedByUser()); ?></td>
```

alfredoct96 replied:
I didn't notice it, thanks!

escopecz commented on app/bundles/EmailBundle/Views/Email/list.html.php:
```diff
@@ -217,6 +250,9 @@
                         <?php echo $view['content']->getCustomContent('email.stats', $mauticTemplateVars); ?>
                         <?php echo $view['content']->getCustomContent('email.stats.below', $mauticTemplateVars); ?>
                     </td>
+                    <td class="visible-lg"><?php echo $item->getDateAdded() ? $view['date']->toFullConcat($item->getDateAdded()) : ''; ?></td>
+                    <td class="visible-lg"><?php echo $item->getDateModified() ? $view['date']->toFullConcat($item->getDateModified()) : ''; ?></td>
```
How about to use the the date format and show the full date and time only on hover?
```suggestion
                    <td class="visible-lg" title="<?php echo $item->getDateAdded() ? $view['date']->toFullConcat($item->getDateAdded()) : ''; ?>">
                        <?php echo $item->getDateAdded() ? $view['date']->toDate($item->getDateAdded()) : ''; ?>
                    </td>
                    <td class="visible-lg" title="<?php echo $item->getDateModified() ? $view['date']->toFullConcat($item->getDateModified()) : ''; ?>">
                        <?php echo $item->getDateModified() ? $view['date']->toDate($item->getDateModified()) : ''; ?>
                    </td>
```
It would save the space, showed only the date as that's the most important part and contain the full information if needed at the same time.

alfredoct96 replied:
Yes, this is better

escopecz commented on app/bundles/EmailBundle/Tests/Controller/EmailControllerFunctionalTest.php:
```diff
@@ -33,6 +33,26 @@ public function setUp(): void
+     */
+    public function testIndexActionWhenNotFiltered(): void
+    {
+        $this->client->request('GET', '/s/emails');
+        $clientResponse = $this->client->getResponse();
+        $this->assertSame(200, $clientResponse->getStatusCode(), 'Return code must be 200.');
```
Tests like these are definitely good to have, but it has little to do with your code change. Only that it's not throwing any exceptions when there are no emails in the list. Can we actually create some emails and test that it shows the data you added to the table? Something like
```suggestion
        $email = $this->createEmail('ABC', 'template');
        $email->setDateAdded(new \DateTime('2020-02-07 20:29:02'));
        $this->em->persist($email);
        $this->em->flush();
        $this->em->clear();

        $this->client->request('GET', '/s/emails');
        $clientResponse = $this->client->getResponse();
        $this->assertSame(200, $clientResponse->getStatusCode(), 'Return code must be 200.');
        $this->assertStringContainsString('2020-02-07 20:29:02', $clientResponse->getContent());
```
⬆️  is untested and will fail as there will be different date time format and we have to test also for the user and date modified but you get the idea.

alfredoct96 replied:
Done!

escopecz reviewed (changes requested):
Thanks! Looks good. I have some suggestions. Please take a look.

alfredoct96 committed:
refactor(list.html.php): added escape and show full date only on hover

alfredoct96 committed:
test(emailcontrollerfunctionaltest): modified a test

now test if email contains saved fields

alfredoct96 commented:
> Thanks! Looks good. I have some suggestions. Please take a look.

Thanks for suggestions! I resolved all points

escopecz commented:
Could you also set the default order by date modified as discussed in https://github.com/mautic/mautic/pull/9690#pullrequestreview-866976865? Ideally all tables for all entities should look and behave similarly.

alfredoct96 committed:
feat(emailcontroller): changed default order to date modified

alfredoct96 commented:
> Could you also set the default order by date modified as discussed in [#9690 (review)](https://github.com/mautic/mautic/pull/9690#pullrequestreview-866976865)? Ideally all tables for all entities should look and behave similarly.

Done @escopecz

escopecz reviewed (dismissed):
Thanks! 👍

kuzmany reviewed (changes requested):
Ordering not work properly, because modified date is empty after create email. Can you check it?

![image](https://user-images.githubusercontent.com/462477/153569542-88101ea2-ba9d-4bd9-ad15-1911fa527e7a.png)

escopecz commented:
@kuzmany that is normal Mautic behavior. We set the date modified only on modification, not on creation. We have a PR that is changing that for all entities but it is blocked by [other PRs](https://github.com/mautic/mautic/pull/7811) and by a chain of 5 other PRs that are blocked by it.

alfredoct96 committed:
refactor(emaillist): added default order to datemodified

alfredoct96 commented:
> Ordering not work properly, because modified date is empty after create email. Can you check it?
> 
> ![image](https://user-images.githubusercontent.com/462477/153569542-88101ea2-ba9d-4bd9-ad15-1911fa527e7a.png)

I added the default order to date modified in the view @kuzmany

kuzmany commented:
@escopecz I understand. But in other parts it works. Now If I create new email, it's in the bottom of list  - in  this case X4 email.

![image](https://user-images.githubusercontent.com/462477/153585330-d3db2eb7-b41f-456f-98bc-f015d8288577.png)

I understand modified date is null and it's correct. We can add it to controller  like in other related PR

https://github.com/mautic/mautic/pull/9690/files#diff-5425f47e560568f92ec341866e7ccfc076b86168f879515feb737a3897ce351eR186 

Probably even better add setDateAdded and setDateModified to construct of entity to cover all ways (controller, api, plugins)

https://github.com/mautic/mautic/blob/5c4bd397c111faebf4f0134670f8076ab5823426/app/bundles/LeadBundle/Entity/LeadEventLog.php#L83

escopecz commented:
We fixed it in https://github.com/mautic/mautic/blob/4.x/app/bundles/CoreBundle/Model/FormModel.php#L231 and prepared migration like 
```php
<?php

declare(strict_types=1);

namespace Mautic\Migrations;

use Doctrine\DBAL\Schema\Schema;
use Mautic\CoreBundle\Doctrine\AbstractMauticMigration;

final class Version20211020142629 extends AbstractMauticMigration
{
    public function up(Schema $schema): void
    {
        $this->addSql("UPDATE {$this->getPrefixedTableName('leads')} SET date_modified = date_added WHERE date_modified IS NULL;");
    }
}
```
For the entities we cared about. I cannot simply cherry-pick it as it is blocked by many other PRs (different base)

kuzmany commented:
@escopecz Ok. But still If I create new email is in the bottom of the list. Even If I expect like in others entities see on the top of the list. Are we okay with it at the moment?

escopecz commented:
I think we should ensure that the date modified is filled in first. Otherwise this PR won't help with UX.

kuzmany commented:
@alfredoct96 are you able to add modified date for new emails?

alfredoct96 committed:
feat(email): set added date and modified date in Email construct

alfredoct96 committed:
feat(emaillist): changed default order to desc

alfredoct96 commented:
> @alfredoct96 are you able to add modified date for new emails?

@kuzmany I pushed the changes, i added created date and modified date in Email entity construct, do you agree with me?

kuzmany commented:
@alfredoct96 

> This branch is out-of-date with the base branch

alfredoct96 commented:
> @alfredoct96
> 
> > This branch is out-of-date with the base branch

Done

kuzmany commented:
@alfredoct96 

> This branch is out-of-date with the base branch

alfredoct96 commented:
> @alfredoct96
> 
> > This branch is out-of-date with the base branch

Done
